clippy_lints/src/append_instead_of_extend.rs

```diff
@@ -0,0 +1,69 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item};
+use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// **What it does:** finds occurrences of `vec.extend(other_vec.drain(..))
+    ///
+    /// **Why is this bad?** Slightly more succinct code and faster code
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // example code where clippy issues a warning
+    /// vec![].extend(vec![1, 2, 3].drain(..))
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// vec![].append(&mut vec![])
+    /// ```
+    pub APPEND_INSTEAD_OF_EXTEND,
+    perf,
+    "default lint description"
+}
+
+declare_lint_pass!(AppendInsteadOfExtend => [APPEND_INSTEAD_OF_EXTEND]);
+
+impl<'tcx> LateLintPass<'tcx> for AppendInsteadOfExtend {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        //validate target object
+        if let ExprKind::MethodCall(method_name, _, exprs, _) = expr.kind {
+            //check source object
+            if let ExprKind::MethodCall(src_method, _, src_args, _) = &exprs[1].kind {
```

Comment:
I reverted from using the if_chain because I could not print out items to see their structure. I can always every back. And my test all pass though some other test failed. Wasn't sure why. Thanks for pointing that out.

Reply:
Oh, thanks for this 👍🏾 I'll definitely use it.